src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1436,16 +1560,23 @@ public bool CleanCacheAndDisposeIfUnused()
                 // Get the current time.  This is compared against each connection's last returned
                 // time to determine whether a connection is too old and should be closed.
                 long nowTicks = Environment.TickCount64;
-                Http2Connection? http2Connection = _http2Connection;
+                Http2Connection[]? localHttp2Connections = _http2Connections;
 
-                if (http2Connection != null)
+                if (localHttp2Connections != null)
                 {
-                    if (http2Connection.IsExpired(nowTicks, pooledConnectionLifetime, pooledConnectionIdleTimeout))
+                    for (int i = 0; i < localHttp2Connections.Length; i++)
                     {
-                        http2Connection.Dispose();
-                        // We can set _http2Connection directly while holding lock instead of calling InvalidateHttp2Connection().
-                        _http2Connection = null;
+                        Http2Connection http2Connection = localHttp2Connections[i];
+                        if (http2Connection != null)
+                        {
+                            if (http2Connection.IsExpired(nowTicks, pooledConnectionLifetime, pooledConnectionIdleTimeout))
+                            {
+                                http2Connection.Dispose();
+                            }
+                        }
                     }
+                    // We can set _http2Connections directly while holding lock instead of calling InvalidateHttp2Connection().
+                    _http2Connections = null;
```

Comment:
This seems wrong... it appears we are discarding all HTTP2 connections here, when what we really want to do is discard just the expired connections.

Reply:
Agree, I also stumbled upon this, but this is existing logic that I just adapted to having a collection instead of a single instance, so I'd prefer to change it in a different PR for a higher visibility of an impact.

Reply:
Now I got your point. Yes, it was definitely a bug. Fixed.